pkg/controllers/localbuild/nginx.go

```diff
@@ -0,0 +1,134 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	nginxNamespace  string = "ingress-nginx"
+	nginxDeployment string = "ingress-nginx-controller"
+)
+
+//go:embed resources/nginx/k8s/*
+var installNginxFS embed.FS
+var timeout = time.After(5 * time.Minute)
+
+func RawNginxInstallResources() ([][]byte, error) {
+	return util.ConvertFSToBytes(installNginxFS, "resources/nginx/k8s")
+}
+
+func NginxInstallResources(scheme *runtime.Scheme) ([]client.Object, error) {
+	rawResources, err := RawNginxInstallResources()
+	if err != nil {
+		return nil, err
+	}
+
+	return k8s.ConvertRawResourcesToObjects(scheme, rawResources)
+}
+
+func newNginxNamespace() *corev1.Namespace {
+	return &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nginxNamespace,
+		},
+	}
+}
+
+func (r LocalbuildReconciler) ReconcileNginx(ctx context.Context, req ctrl.Request, resource *v1alpha1.Localbuild) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	nginxNSClient := client.NewNamespacedClient(r.Client, nginxNamespace)
+	installObjs, err := NginxInstallResources(r.Scheme)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Ensure namespace exists
+	nginxNewNS := newNginxNamespace()
+	if err = r.Client.Get(ctx, types.NamespacedName{Name: nginxNamespace}, nginxNewNS); err != nil {
+		// We got an error so try creating the NS
+		if err = r.Client.Create(ctx, nginxNewNS); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	log.Info("Installing Nginx resources")
+	for _, obj := range installObjs {
+		if obj.GetObjectKind().GroupVersionKind().Kind == "Deployment" {
+			switch obj.GetName() {
+			case nginxDeployment:
+				gotObj := appsv1.Deployment{}
+				if err := r.Client.Get(ctx, types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}, &gotObj); err != nil {
+					if err = controllerutil.SetControllerReference(resource, obj, r.Scheme); err != nil {
+						log.Error(err, "Setting controller reference for Nginx deployment", "deployment", obj)
+						return ctrl.Result{}, err
+					}
+				}
+			}
+		}
+
+		// Create object
+		if err = k8s.EnsureObject(ctx, nginxNSClient, obj, nginxNamespace); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Wait for Nginx to become available
+	ready := make(chan error)
+	go func([]client.Object) {
+		for {
+			for _, obj := range installObjs {
+				if obj.GetObjectKind().GroupVersionKind().Kind == "Deployment" {
+					switch obj.GetName() {
+					case nginxDeployment:
+						gotObj := appsv1.Deployment{}
+						if err := r.Client.Get(ctx, types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}, &gotObj); err != nil {
+							ready <- err
+							return
+						}
+
+						if gotObj.Status.AvailableReplicas >= 1 {
+							close(ready)
+							return
+						}
+					}
+				}
+			}
+
+			time.Sleep(1 * time.Second)
+		}
+	}(installObjs)
+
+	select {
+	case <-timeout:
```

Comment:
I am not well versed enough to know if this is the idiomatic way of waiting for the resource to become available. Isn't it also possible to requeue like so: https://github.com/cnoe-io/idpbuilder/blob/1440d25e8d74ba44fdc31a2818fa0efaf4e6b8a8/pkg/controllers/localbuild/controller.go#L167

Reply:
interesting. didnt know about this. let me look into it.

Reply:
ok yeah so I did dig and `RequeueAfter` only schedules to rerun the reconciliation loop. The check we have in the code is to make sure `Nginx.Status.AvailableReplicas > 0` so we can have the localbuild status for Nginx readiness set to true.